documentation/projects/proposals/popularity_optimizations/20230420-implementation_plan_popularity_optimizations.md

```diff
@@ -0,0 +1,488 @@
+matches what the ingestion server already expects, we can simply change the
+table name each place it is referenced:
+
+- [`_get_shared_cols`](https://github.com/WordPress/openverse/blob/003c0d3e5c2813b671120144756659775f3369f4/ingestion_server/ingestion_server/ingest.py#L75)
+- [`refresh_api_table`](https://github.com/WordPress/openverse/blob/003c0d3e5c2813b671120144756659775f3369f4/ingestion_server/ingestion_server/ingest.py#L310)
+- [`get_copy_data_query`](https://github.com/WordPress/openverse/blob/003c0d3e5c2813b671120144756659775f3369f4/ingestion_server/ingestion_server/queries.py#L206)
+
+## Dependencies
+
+### Tools & packages
+
+<!-- Describe any tools or packages which this work might be dependent on. If multiple options are available, try to list as many as are reasonable with your own recommendation. -->
+
+We should not need any additional dependencies.
+
+### Other projects or work
+
+<!-- Note any projects this plan is dependent on. -->
+
+This does not depend on any existing projects.
+
+### Infrastructure
+
+<!-- Note any infrastructure this plan is dependent on. -->
+
+This project directly affects our Catalog database schema, Airflow DAGs, and
+ingestion server. All necessary changes are described in this document.
+
+Notably, because this plan would allow us to drop the materialized `image_view`
+and `audio_view`, we should net some DB storage savings in the long term.
+
+## Alternatives
+
+### Duplicate table
+
+Another approach was considered which, rather than updating records directly on
+the `image` table, would have instead created a duplicate table with updated
+scores and ultimately swapped tables. This has the benefit of avoiding
+concurrent writes to the `image` table, but the following disadvantages:
+
+- While the duplicate table is being built, writes to the `image` table from
+  provider DAGs would need to be propagated to it via a Postgres
+  [`Trigger`](https://www.postgresql.org/docs/current/plpgsql-trigger.html)
+  rule. Thus the same deadlocking precautions are necessary, and the writes must
+  still be batched, offering no improved performance.
+- This approach requires much more space for the duplicate table (although
+  roughly equivalent to the current materialized view approach).
+- This approach is much less efficient because it unnecessarily re-inserts
+  **all** rows, including those that do not support popularity data at all.
+
+### Separate `image_popularity` view
+
+We considered keeping the `image` table as is, and an `image_popularity`
+materialized view equivalent to the current `image_view`. A new,
+non-materialized `image_view` would join `image` and `image_popularity` and be
+used for the data refresh. The advantage is that the `image_popularity` view
+could be refreshed independently of the data refresh; this approach would also
+be very quick and easy to implement. The disadvantages:
+
+- Would require an expensive `join` at the time of data refresh
```

Comment:
I was thinking on this more, and I actually wonder if it might be possible to combine the proposed at-ingestion-calculation approach with having a separate table for popularity data. We could have a fully separate table for `image_popularity` (_not_ a matview, but full table) that gets updated on ingestion and also updated by the `popularity_refresh` DAG. This has the advantage of preventing our image table from growing wider, and the join can happen by way of a non-materialized `image_view` which joins the two. That join would only be done when the table is queried, i.e. during the data refresh. It might also mean that updates to the table by both the ingestion and the popularity refresh could be quicker, since the table is only two columns wide. And we'd only be storing popularity data for records we have, rather than having a null column for all our data without popularity info. On the flip side, it'd bring some additional complexity at ingestion because we'd need a way to write to both tables (and I believe the `identifier` which is used as our ID for a record is generated on insert into the primary media table) 🤔 I'm also not sure it gets us out of the concurrent write issue you've described. What do you think? Are there other drawbacks or benefits that I might have missed in thinking about this?

Reply:
> It might also mean that updates to the table by both the ingestion and the popularity refresh could be quicker, since the table is only two columns wide.

😮 Is an _update_ of a single column necessarily faster if the table it's in is smaller? Wow!

>  And we'd only be storing popularity data for records we have, rather than having a null column for all our data without popularity info.

This definitely strikes me as an advantage, although it would be a theoretically diminishing one if we're able to add popularity data to a larger proportion of our providers.

On the other hand, each row would also have to duplicate the `identifier` in order to join the tables, so if the proportion of records that do have pop data increases, we might end up using _more_ space. The proportion of records with pop data is actually already quite high (I believe >50%), because all the Flickr records have it.

> On the flip side, it'd bring some additional complexity at ingestion because we'd need a way to write to both tables (and I believe the identifier which is used as our ID for a record is generated on insert into the primary media table) 🤔 I'm also not sure it gets us out of the concurrent write issue you've described.

I'll have to double check but I'm actually fairly certain that `identifier` is generated and present in the tsvs, so I _think_ we have it before insert in the primary media table. We would probably still need to add a separate task for writing the scores to a separate table though, which would add a little bit of complexity if we still want all the values in the same tsv.

I don't think it prevents the concurrent write issues, since we'd still have multiple DAGs writing to the same table. Thinking off the cuff, if the popularity writes are in a separate task in the provider DAG, we could add a sensor that skips that task if the `refresh_<provider>_popularity` task is running. But I don't think that will work, because the provider DAG might be adding _new_ records. So unfortunately I don't think the popularity refresh tasks would be any more performant, unless the individual `update`s run faster on smaller tables.

> Are there other drawbacks or benefits that I might have missed in thinking about this?

I think it would be interesting to benchmark copying directly from the image table vs copying from a non-materialized view with the `join`. Surely there must be some performance impact, but it might be totally insignificant. In either case I don't see how it could be _faster._

I think my gut feeling is that this would not be easier to implement (and might add extra complexity), and likely wouldn't improve the performance of the popularity refresh, but _might_ make the data refresh a little slower. I'm really intrigued by the idea that the smaller table might make popularity updates a little bit faster, but the speed of the popularity _refresh_ is theoretically the least important aspect of the system.

Reply:
> I'll have to double check but I'm actually fairly certain that identifier is generated and present in the tsvs, so I think we have it before insert in the primary media table. We would probably still need to add a separate task for writing the scores to a separate table though, which would add a little bit of complexity if we still want all the values in the same tsv.

Unfortunately we don't have `identifier` before inserting the media from the loading table into the main table:
https://github.com/WordPress/openverse/blob/59a4411e4363207627e54467f7cf6ad40b1a4b0b/catalog/dags/common/storage/db_columns.py#L10-L11

>😮 Is an update of a single column necessarily faster if the table it's in is smaller? Wow!
When doing updates, Postgres writes an entirely new row, unless it is a [HOT update](https://www.cybertec-postgresql.com/en/hot-updates-in-postgresql-for-better-performance/). I learned this when trying to update some value in the upstream database and it would take more than 40 days :)

Reply:
Oof, thanks for double checking that Olga. That would be an additional complication, then :/ We could possibly change where the `identifier`s are generated so we _do_ have them in the tsvs. We'd definitely want to avoid extra queries.

Reply:
If we create identifiers before writing to TSVs, we will have inconsistencies with the database. If the identifier is created during ingestion and written to TSV, then during the data refresh, for items that we already have in the database, we would replace the identifier from the TSV with the identifier from the database :/

Reply:
Is it possible to use Python's UUID module to generate the identifier and only have Postgres generate one when one isn't provided in the TSV? Apologies if this misses an obvious piece of the puzzle of how ingested records get written. Basically though my question is: can the UUID be generated in Python when the record is retrieved? Is there an issue with needing to confirm that records that are being ingested aren't duplicates of existing records, in which case they'd already have a UUID (queryable via the URL)? Would a Redis cache of URL -> UUID for existing records help with that?

Reply:
Currently, we collect all the data from the provider using the providerDataIngester, and write it all into a TSV. Then we load the TSV into a loading table and clean it up (remove duplicates on foreign_id and provider pair, and records with null for required fields). Then we upsert these rows into the main `image` or `audio` table. During the upsert, if there's already an item with the same `url`, we update it. Otherwise, we create a insert a new row and create a new UUID then.

We could do something like `SELECT identifier FROM image where url='image_url'` during ingestion, but I have no idea how that would affect the speed or performance. Also, if we create a UUID during the ingestion, we will have to make sure that there is no collision of UUIDs created by different provider scripts running at the same time.

Reply:
> Also, if we create a UUID during the ingestion, we will have to make sure that there is no collision of UUIDs created by different provider scripts running at the same time.

This is theoretically guaranteed to be a non-issue with UUIDs right? [Wikipedia claims that you need to generate 2.71 quintillion UUIDs before the chance of collision is 50%](https://en.wikipedia.org/wiki/Universally_unique_identifier#Collisions). Granted we'd need to make sure collisions don't happen, but it's sure to be unlikely. Even with our several hundred million records, the chances are miniscule. From the article linked above:

> the probability to find a duplicate within 103 trillion version-4 UUIDs is one in a billion.

I'd venture to say that the chances are so small we should feel comfortable deferring even considering how to solve that to a later project and date.

> We could do something like SELECT identifier FROM image where url='image_url' during ingestion, but I have no idea how that would affect the speed or performance.

That's along the lines of what I was thinking should be possible. [If `url` is indexed](https://github.com/WordPress/openverse/blob/697f62f01a32cb7fcf2f4a7627650a113cba40da/catalog/dags/common/loader/sql.py#L105) it should be very fast, right?

Reply:
Theoretically I agree with you, @sarayourfriend. But I'd still like to hear what others think, because I don't have enough experience with it.

Reply:
@sarayourfriend's proposal is what I had in mind, yep. I think that would make it possible to generate the identifiers in advance, so we could keep everything in one tsv and eliminate the need for additional queries. In the incredibly unlikely event of a collision the provider DAG would fail on the load step, which can be easily retried, so there is no risk as far as I can see.

That being said, this would still add complexity, and I'm not sure I see any equivalent _advantage_ to this approach over the proposed one. Arguably storing the calculated popularity scores in another table still violates normalization principles. I also think it would end up taking up more space and potentially make the data refresh take longer.

I'd like to move this IP into the revision stage, but I want to give @AetherUnbound especially a chance to respond to the thread here.

Reply:
Apologies for the delay in getting back to this - I've been mulling this over and I don't have a strong argument for why it should be in a separate table. My _gut_ says that having it in a separate table is "the right way" to do it and would have a positive impact on various optimizations, but I can't put much more behind that than a feeling of not wanting our primary tables to get larger. All that to say, I don't have a good reason for this being a blocking objection, so we can disregard it and stick with the plan as proposed 🙂 No other blockers on moving into the objection stage on my end!